In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.metrics import accuracy_score

In [2]:
mail = pd.read_csv('mail_data.csv')

In [5]:
mail.isnull().sum()

,0
Category,0
Message,0


In [6]:
mail_data = mail.where((pd.notnull(mail)),'') #replace the null values with the null string

In [11]:
mail_data.head()

,Category,Message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [12]:
mail_data.tail()

,Category,Message
5567,spam,This is the 2nd time we have tried 2 contact u...
5568,ham,Will ü b going to esplanade fr home?
5569,ham,"Pity, * was in mood for that. So...any other s..."
5570,ham,The guy did some bitching but I acted like i'd...
5571,ham,Rofl. Its true to its name


In [13]:
#checking the number of rows and cols
mail_data.shape  #5572 rows and 2 cols

(5572, 2)

In [15]:
#label encoding

#label spam email as 0 and ham email as 1

mail_data.loc[mail_data['Category'] == 'spam', 'Category',] = 0
mail_data.loc[mail_data['Category'] == 'ham', 'Category',] = 1

#spam - 0(fraud or spam mail) and ham - 1(actual mail)

In [16]:
x = mail_data['Message']
y = mail_data['Category']

In [17]:
print(x)

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                 Will ü b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: Message, Length: 5572, dtype: object


In [18]:
print(y)

0       1
1       1
2       0
3       1
4       1
       ..
5567    0
5568    1
5569    1
5570    1
5571    1
Name: Category, Length: 5572, dtype: object


In [19]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=3)

In [20]:
print(x_train)

3075                  Don know. I did't msg him recently.
1787    Do you know why god created gap between your f...
1614                         Thnx dude. u guys out 2nite?
4304                                      Yup i'm free...
3266    44 7732584351, Do you want a New Nokia 3510i c...
                              ...                        
789     5 Free Top Polyphonic Tones call 087018728737,...
968     What do u want when i come back?.a beautiful n...
1667    Guess who spent all last night phasing in and ...
3321    Eh sorry leh... I din c ur msg. Not sad alread...
1688    Free Top ringtone -sub to weekly ringtone-get ...
Name: Message, Length: 4457, dtype: object


In [21]:
print(y_train)

3075    1
1787    1
1614    1
4304    1
3266    0
       ..
789     0
968     1
1667    1
3321    1
1688    0
Name: Category, Length: 4457, dtype: object


In [22]:
print(x_test)

2632    URGENT! Your mobile No 077xxx WON a £2,000 Bon...
454     Ok i will tell her to stay out. Yeah its been ...
983     Congrats! 2 mobile 3G Videophones R yours. cal...
1282        Am I the only one who doesn't stalk profiles?
4610                               Y de asking like this.
                              ...                        
4827                        Haha, just what I was thinkin
5291      Xy trying smth now. U eat already? We havent...
3325    I don wake since. I checked that stuff and saw...
3561    Lol I know! Hey someone did a great inpersonat...
1136                      K do I need a login or anything
Name: Message, Length: 1115, dtype: object


In [23]:
print(y_test)

2632    0
454     1
983     0
1282    1
4610    1
       ..
4827    1
5291    1
3325    1
3561    1
1136    1
Name: Category, Length: 1115, dtype: object


In [28]:
from sklearn.feature_extraction.text import TfidfVectorizer

#transform the text data to feature vectors that can be used as input to the logistic regression
features_extraction = TfidfVectorizer(min_df=1, stop_words='english', lowercase=True)
x_train_features = features_extraction.fit_transform(x_train)
x_test_features = features_extraction.transform(x_test)

In [29]:
#convert y_train and y_test to numbers

y_train = y_train.astype('int')
y_test = y_test.astype('int')


In [30]:
print(x_train_features)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 34775 stored elements and shape (4457, 7431)>
  Coords	Values
  (0, 2329)	0.38783870336935383
  (0, 3811)	0.34780165336891333
  (0, 2224)	0.413103377943378
  (0, 4456)	0.4168658090846482
  (0, 5413)	0.6198254967574347
  (1, 3811)	0.17419952275504033
  (1, 3046)	0.2503712792613518
  (1, 1991)	0.33036995955537024
  (1, 2956)	0.33036995955537024
  (1, 2758)	0.3226407885943799
  (1, 1839)	0.2784903590561455
  (1, 918)	0.22871581159877646
  (1, 2746)	0.3398297002864083
  (1, 2957)	0.3398297002864083
  (1, 3325)	0.31610586766078863
  (1, 3185)	0.29694482957694585
  (1, 4080)	0.18880584110891163
  (2, 6601)	0.6056811524587518
  (2, 2404)	0.45287711070606745
  (2, 3156)	0.4107239318312698
  (2, 407)	0.509272536051008
  (3, 7414)	0.8100020912469564
  (3, 2870)	0.5864269879324768
  (4, 2870)	0.41872147309323743
  (4, 487)	0.2899118421746198
  :	:
  (4454, 2855)	0.47210665083641806
  (4454, 2246)	0.47210665083641806
  (4455, 4456)	0.24

In [31]:
print(x_test_features)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 7687 stored elements and shape (1115, 7431)>
  Coords	Values
  (0, 1)	0.2381316303003606
  (0, 9)	0.2852706805264544
  (0, 14)	0.26797874471323896
  (0, 20)	0.30668032384591537
  (0, 306)	0.23975986557206702
  (0, 405)	0.2381316303003606
  (0, 1041)	0.28016206931555726
  (0, 1082)	0.2451068436245027
  (0, 1361)	0.25132445289897426
  (0, 1405)	0.3176863938914351
  (0, 1549)	0.2646498848307188
  (0, 4386)	0.18353336340308998
  (0, 5213)	0.1988547357502182
  (0, 5373)	0.2365698724638063
  (0, 6920)	0.20571591693537986
  (0, 7271)	0.1940327008179069
  (1, 3491)	0.496093956101028
  (1, 4418)	0.3457696891316818
  (1, 4729)	0.22965776503163893
  (1, 6214)	0.3621564482127515
  (1, 6507)	0.26731535902873493
  (1, 6588)	0.3298937975962767
  (1, 6732)	0.42473488678029325
  (1, 7368)	0.29957800964520975
  (2, 201)	0.2824102268489399
  :	:
  (1110, 6591)	0.7327660015422193
  (1111, 2440)	0.4137350055985486
  (1111, 3227)	0.44384935772735

In [32]:
#train the model

model = LogisticRegression()

In [34]:
model_train = model.fit(x_train_features, y_train)

In [40]:
from sklearn.metrics import accuracy_score

pred_train = model.predict(x_train_features)
accuracy_train = accuracy_score(y_train, pred_train)

In [41]:
print('accuracy on training data: ', accuracy_train)

accuracy on training data:  0.9676912721561588


In [46]:
from sklearn.metrics import accuracy_score
pred_test = model.predict(x_test_features)
accuracy_test = accuracy_score(y_test,pred_test)

In [47]:
print('accuracy on test data: ', accuracy_test)

accuracy on test data:  0.9668161434977578


In [61]:
input_mail =["	WINNER!! As a valued network customer you have been selected to receivea £900 prize reward! To claim call 09061701461. Claim code KL341. Valid 12 hours only."]

In [62]:
#convert text to features

input_features = features_extraction.transform(input_mail)

#making prediction

prediction = model.predict(input_features)
print(prediction)

if (prediction[0]==1):
    print('Ham mail')
else:
    print('Spam mail')

[0]
Spam mail


# Spam/Ham Email Classifier

## Project Overview
This project implements a machine learning model to classify emails as either 'spam' or 'ham' (legitimate email). The goal is to build a robust classifier that can distinguish between unwanted promotional or malicious emails and important personal or professional correspondence.

## Dataset
The model was trained and evaluated using the `mail_data.csv` dataset, which contains email messages labeled as 'spam' or 'ham'.

## Methodology

### 1. Data Preprocessing
*   Missing values were handled by replacing them with empty strings.
*   Email categories ('spam' and 'ham') were converted into numerical labels (0 for spam, 1 for ham) using label encoding.

### 2. Feature Extraction
*   The text data from the email messages was transformed into numerical feature vectors using **TF-IDF Vectorization**. This technique assigns weights to words based on their frequency in a document and across the entire corpus, highlighting words that are important in a specific email but not common everywhere.

### 3. Model Training
*   A **Logistic Regression** model was chosen for classification due to its effectiveness in binary classification tasks and its interpretability.
*   The data was split into training and testing sets (80% training, 20% testing) to evaluate the model's generalization capability.

## Performance

The trained Logistic Regression model achieved the following accuracies:
*   **Accuracy on Training Data**: 0.9677 (or approximately 96.77%)
*   **Accuracy on Test Data**: 0.9668 (or approximately 96.68%)

These high accuracy scores indicate that the model performs well in distinguishing between spam and ham emails on both seen and unseen data.

## How to Use (Prediction Example)

To predict whether a new email is spam or ham, you can use the trained `features_extraction` (TF-IDF Vectorizer) and `model` (Logistic Regression Classifier):

```python
# Example input email
input_mail = ["Congratulations! You've won a free iPhone. Click here to claim your prize!"]

# Convert the input text to feature vectors using the trained TF-IDF vectorizer
input_features = features_extraction.transform(input_mail)

# Make a prediction using the trained Logistic Regression model
prediction = model.predict(input_features)

# Interpret and print the prediction
if (prediction[0] == 1):
    print('Ham mail')
else:
    print('Spam mail')
```

This will output whether the `input_mail` is classified as 'Ham mail' or 'Spam mail'.